### ISSUE TITLE
Checksum assertion failure in disclaim_test

### ISSUE THREAD

**Issue by ivmai** [ivmai/bdwgc#400]
Build: https://app.travis-ci.com/github/ivmai/bdwgc/jobs/550649361
Source: master (1f1e0df060)
Host: Linux/ppc64le
Config: CMAKE_CONFIG="Release" CMAKE_OPTIONS="-Dbuild_tests=ON -Denable_cplusplus=ON -Denable_gc_assertions=ON"

Output:
Test command: /home/travis/build/ivmai/bdwgc/disclaim_test
16: Threaded disclaim test.
16: Assertion failure, line 127: p->checksum == checksum

**Comment by ivmai** [comment]
Probably related failure: https://app.travis-ci.com/github/ivmai/bdwgc/jobs/555921895
FAIL: disclaim_weakmap_test
Source: master (742febb2f)

**Comment by ivmai** [comment]
@paurkedal, 
Could this be because of a lack of threads synchronization?
checksum is set in pair_new() but there's no barrier on object construction (main or some created thread);
checksum is verified in pair_dct() which is called from a collector thread.

**Comment by paurkedal** [comment]
That sounds like a plausible explanation. That means other fields could have the same issue, as well.  Would it be possible to add a common fence just before the return?

**Comment by ivmai** [comment]
Build: https://app.travis-ci.com/github/ivmai/bdwgc/jobs/570835935
Source: master (107cfe0)
Output (disclaimtest):
Assertion failure, line 110: cd == (void *)PTR_HASH(p)

**Comment by ivmai** [comment]
> That sounds like a plausible explanation. That means other fields could have the same issue, as well. Would it be possible to add a common fence just before the return?

Yes, I will check it in disclaimtest. (weakmaptest already contains unlock but probably weakmaptest has failure reason other than that for disclaimtest.)

**Comment by ivmai** [comment]
Thinking about the issue more: pair_dct is called during reclaim phase for objects which have been marked as garbage, garbage collection is performed with the world stopped, on Linux each thread calls sem_post and the collecting thread calls sem_wait: so it looks like the barrier already exists (sem_post -> sem_wait).

**Comment by ivmai** [comment]
https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap04.html#tag_04_11 : 
sem_post and sem_wait synchronize memory with respect to other threads

**Comment by ivmai** [comment]
So, right now I don't have any idea of the root cause of the failures. Need to reproduce the failure on some ppc64le host (e.g. on the GCC farm).

**Comment by ivmai** [comment]
Probably related: SIGSEGV in disclaim_weakmap_test.
Source: release-8_2 (cd1fbc1d)
Host: Linux/ppc64le:
Link: https://app.travis-ci.com/github/ivmai/bdwgc/jobs/581044063 (rebuilt)
